pantry/src/pantry.rs

```diff
@@ -0,0 +1,431 @@
+                )
+                .send()
+                .await?;
+
+            let content_length = response
+                .headers()
+                .get(reqwest::header::CONTENT_LENGTH)
+                .ok_or("no content length!")
+                .map_err(|e| anyhow!(e))?;
+
+            let content_length = usize::from_str(content_length.to_str()?)?;
+
+            assert_eq!(content_length, end - start);
+            assert!(content_length <= Self::MAX_CHUNK_SIZE);
+            assert!(content_length % volume_block_size as usize == 0);
+
+            let bytes = response.bytes().await?;
+
+            if let Some(ref mut hasher) = hasher {
+                hasher.update(&bytes);
+            }
+
+            self.volume
+                .write_to_byte_offset(start as u64, bytes)
+                .await?;
+        }
+
+        // flush
+
+        self.volume.flush(None).await?;
+
+        if let Some(hasher) = hasher {
+            let digest = hex::encode(hasher.finalize());
+
+            match expected_digest.unwrap() {
+                ExpectedDigest::Sha256(expected_digest) => {
+                    if expected_digest != digest {
+                        bail!(
+                            "sha256 digest mismatch! expected {}, saw {}",
+                            expected_digest,
+                            digest,
+                        );
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn snapshot(&self, snapshot_id: String) -> Result<()> {
+        self.volume
+            .flush(Some(SnapshotDetails {
+                snapshot_name: snapshot_id,
+            }))
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn bulk_write(&self, offset: u64, data: Vec<u8>) -> Result<()> {
+        if data.len() > Self::MAX_CHUNK_SIZE {
+            bail!(
+                "data len {} over max chunk size{}!",
+                data.len(),
+                Self::MAX_CHUNK_SIZE
+            );
+        }
+
+        self.volume
+            .write_to_byte_offset(offset, data.into())
+            .await?;
+
+        Ok(())
+    }
+
+    pub async fn scrub(&self, log: &Logger) -> Result<()> {
+        self.volume.scrub(log).await?;
+        Ok(())
+    }
+
+    pub async fn detach(&self) -> Result<()> {
+        self.volume.flush(None).await?;
+        self.volume.deactivate().await?;
+        Ok(())
+    }
+}
+
+/// Pantry stores opened Volumes in-memory
+pub struct Pantry {
+    pub log: Logger,
+
+    /// Store a Volume Construction Request and Volume, indexed by id. Use this
+    /// Mutex -> Arc<Mutex> structure in order for multiple requests to act on
+    /// multiple PantryEntry objects at the same time.
+    entries: Mutex<BTreeMap<String, Arc<Mutex<PantryEntry>>>>,
+
+    /// Pantry can run background jobs on Volumes, and currently running jobs
+    /// are stored here.
+    jobs: Mutex<BTreeMap<String, JoinHandle<Result<()>>>>,
```

Comment:
There is no connection here between the job `string` (`UUID`?) here and the volume it is running on, right?  Meaning the jobs and entries have no connection.

Is it desirable to have them connected?  To know what jobs are working on what entries?

Reply:
There is currently no connection, right. I decided against connecting them in the pantry because Nexus will be where that connection will be made (as part of a saga's context for example)